src/components/forms/TextInput/TextInput.tsx

```diff
@@ -50,21 +55,83 @@ export const TextInput = ({
       'usa-input--success': isSuccess,
       'usa-input--small': isSmall,
       'usa-input--medium': isMedium,
+      'usa-masked': mask,
     },
     className
   )
 
-  return (
+  const [inputValue, setInputValue] = useState('')
+  const input = (
     <input
       data-testid="textInput"
       className={classes}
       id={id}
       name={name}
       type={type}
       ref={inputRef}
+      maxLength={
+        inputProps.maxLength || !mask ? inputProps.maxLength : mask.length
+      }
+      onChange={mask ? handleChange : inputProps.onChange}
+      value={mask ? inputValue : inputProps.value}
       {...inputProps}
     />
   )
+
+  const [maskValue, setMaskValue] = useState(mask)
+  const [iValue, setIValue] = useState(<i></i>)
+  function handleChange(e: React.ChangeEvent<HTMLInputElement>): void {
+    const value = e.target.value
+    const maskData = charset || mask
+    if (undefined === maskData) return
+    const strippedValue = charset
+      ? value.replace(/\W/g, '')
+      : value.replace(/\D/g, '')
+    const isInteger = (v: string) => !Number.isNaN(parseInt(v, 10))
+    const isLetter = (v: string) => (v ? v.match(/[A-Z]/i) : false)
+    const maskedNumber = '_#dDmMyY9'
+    const maskedLetter = 'A'
+    let newValue = ''
+    for (let m = 0, v = 0; m < maskData.length; m++) {
+      const isInt = isInteger(strippedValue[v])
+      const isLet = isLetter(strippedValue[v])
+      const matchesNumber = maskedNumber.indexOf(maskData[m]) >= 0
+      const matchesLetter = maskedLetter.indexOf(maskData[m]) >= 0
+      if ((matchesNumber && isInt) || (charset && matchesLetter && isLet)) {
+        newValue += strippedValue[v++]
+      } else if (
+        strippedValue[v] === undefined || // if no characters left and the pattern is non-special character
+        (!charset && !isInt && matchesNumber) ||
+        (charset && ((matchesLetter && !isLet) || (matchesNumber && !isInt)))
+      ) {
+        break
+      } else {
+        newValue += maskData[m]
+      }
+    }
+    const newMaskValue = mask ? mask.substring(newValue.length) : ''
+    setMaskValue(newMaskValue)
+    setIValue(<i>{newValue}</i>)
+    setInputValue(newValue)
+    inputProps.onChange
+  }
+
+  if (mask) {
+    return (
+      <span className="usa-input-mask">
+        <span
+          className="usa-input-mask--content"
+          aria-hidden
+          data-testid={`${id}Mask`}>
+          {iValue}
+          {maskValue}
+        </span>
+        {input}
+      </span>
+    )
+  } else {
+    return input
+  }
```

Comment:
**suggestion (not blocking)**: On a quick scan, this looks like a pretty good reason to make InputMask its own component, and have it just render a TextInput in place of `{input}`. Since USWDS considers it a separate component as well, that makes it a bit more compelling

That would keep TextInput's code simple (and likely unchanged, which is good, means no potential for regressions), while leaning on componentization and composition more heavily, which are really good things.

Not gonna block on it because you might have tried that or discovered a reason why that won't work.  If it is possible (or if it might be possible) and we merge this, I'll recommend creating a github issue to split them apart and perform a subsequent breaking release.

Reply:
yeah i also considered this, and my logic was just that this is a feature of text fields, but the points you make are probably more important and probably wouldn't cause too much extra thought on behalf of the user. Lemme quickly refactor this and submit one more commit for review